File: UnityHello/Assets/Game/Scripts/Framework/Locker.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public class Locker<T>
{
    private List<T> mLockKeys = new List<T>();

    public void Lock(T key)
    {
        if (!mLockKeys.Contains(key))
        {
            mLockKeys.Add(key);
        }
    }

    public void UnLock(T key)
    {
        if (mLockKeys.Contains(key))
        {
            mLockKeys.Remove(key);
        }
    }

    public bool IsLock(T key)
    {
        return mLockKeys.Contains(key);
    }

    public bool HasLock()
    {
        return mLockKeys.Count > 0;
    }
}

public class SimpleLocker
{
    private bool mLocked = false;

    public void Lock()
    {
        mLocked = true;
    }

    public void UnLock()
    {
        mLocked = false;
    }

    public bool IsLock()
    {
        return mLocked;
    }
}

File: UnityHello/Assets/Lua/lscripts/Locker.lua
Locker = {}

function Locker:new(o)
	o = o or {}
	setmetatable(o, self)
	self.__index = self

	o.lockKeys = {}

	return o
end

function Locker:lock(lockKey)
	if(self.lockKeys[lockKey] == nil) then
		self.lockKeys[lockKey] = true
	end
end

function Locker:unLock(lockKey)
	if(self.lockKeys[lockKey] ~= nil) then
		self.lockKeys[lockKey] = nil
	end
end

function Locker:isLock(lockKey)
	return self.lockKeys[lockKey] ~= nil
end

function Locker:hasLock()
	for k, v in pairs(self.lockKeys) do
		if(k ~= nil) then
			return true
		end
	end
	return false
end

